### PR TITLE
authz: add built-in roles to the database

### PR DESCRIPTION
**Pull request by davepacheco**
For where this fits into the authz project, see #419.  The Omicron Polar file defines a fixed set of roles.  This change loads them into the database.

There is currently no way I know of to verify that our database roles match the roles in the Polar files, but there is a feature coming from Oso to get this information.

### CODE REVIEW

File: nexus/src/authn/mod.rs [hunk 1]
```diff
@@ -26,93 +26,12 @@
 
 pub mod external;
 
-use lazy_static::lazy_static;
```
Comment (davepacheco) [inline nexus/src/authn/mod.rs:29]: Much of this file got moved over to `nexus/src/db/fixed_data/users.rs`.

File: nexus/src/db/datastore.rs [hunk 2]
```diff
@@ -2154,6 +2155,120 @@ impl DataStore {
...
+    /// List built-in roles
+    pub async fn roles_builtin_list_by_name(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, (String, String)>,
+    ) -> ListResultVec<RoleBuiltin> {
+        use db::schema::role_builtin::dsl;
+        opctx.authorize(authz::Action::ListChildren, authz::FLEET)?;
+
+        // TODO-column This is essentially a multi-column version of
```
Comment (davepacheco) [inline nexus/src/db/datastore.rs:2168]: Re: this comment: I spent a while trying to build a multi-column version of `paginated` but could not get the Diesel trait bounds right.  I gave up on that because there would only be this one caller right now.

File: common/src/sql/dbinit.sql [hunk 3]
```diff
@@ -707,6 +707,56 @@ INSERT INTO omicron.public.user_builtin (
...
+ * * "time_created": not that useful because it's whenever the system was
+ *   initialized, and we have plenty of other timestamps for that
+ * * "time_modified": does not apply because the role cannot be changed
+ * * "time_deleted" does not apply because the role cannot be deleted
+ *
+ * If the set of roles and their permissions are fixed, why store them in the
+ * database at all?  Because what's dynamic is the assignment of roles to users.
+ * We [will] have a separate table that says "user U has role ROLE on resource
+ * RESOURCE".  How do we represent the ROLE part of this association?  We use a
+ * foreign key into this "role_builtin" table.
```
Comment (teisenbe) [inline common/src/sql/dbinit.sql:750]: Since we don't use true foreign keys, and we're not storing any mutable data in the table, I'm not sure if we get much benefit from having this table versus keeping the table in our code. The "foreign key" columns in the user permissions table can just as easily be used to reference into a map in Nexus. Are there benefits we expect from having it in the database?

File: nexus/tests/integration_tests/roles_builtin.rs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    );
+
+    // This endpoint uses a custom pagination scheme that is easy to get wrong.
+    // Let's test that all markers do work.
+    let roles_paginated =
+        NexusRequest::iter_collection_authn(&testctx, "/roles", "", 1)
+            .await
+            .expect("failed to iterate all roles");
+    assert_eq!(roles, roles_paginated.all_items);
+    assert_eq!(roles.len() + 1, roles_paginated.npages);
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/tests/integration_tests/roles_builtin.rs:71]: Sure, will do.

File: nexus/src/external_api/views.rs [hunk 5]
```diff
@@ -198,3 +198,26 @@ impl Into<User> for model::UserBuiltin {
...
+ * ROLES
+ */
+
+/**
+ * Client view of a [`Role`]
+ */
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+pub struct Role {
+    pub name: String,
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/views.rs:212]: Sure, will do.

File: common/src/sql/dbinit.sql [hunk 6]
```diff
@@ -707,6 +707,56 @@ INSERT INTO omicron.public.user_builtin (
...
+ *
+ * Internally, we can treat the tuple (resource type, role name) as a composite
+ * primary key.  Externally, we expose this as the full name.  This is
+ * consistent with RFD 43 and other IAM systems.
+ *
+ * These fields look awfully close to the identity metadata that we use for most
+ * other tables.  But they're just different enough that we can't use most of
+ * the same abstractions:
+ *
+ * * "id": We have no need for a uuid because the (role_name, full_name) is
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply common/src/sql/dbinit.sql:735]: Yes, that's right.  Nice catch.

File: common/src/sql/dbinit.sql (hunk 3 above)
Comment (davepacheco) [reply common/src/sql/dbinit.sql:750]: You're right that the table doesn't buy us a whole lot today and I've been on the fence about whether to even do this.  I initially decided to make this table as part of standard normalization.  In retrospect that seems dubious if the primary key comprises the entire row.  I later added metadata from RFD 43 -- the names of the roles that we'd show in the UI.  Another plus side to doing this now is that I think this schema resembles what we might do in the future with custom roles, and it'd be nice to get some experience with it.  This too feels like a weak reason.

That said, if we _are_ going to have a table of this information, even if it's immutable, I think it's simpler to keep it in the database with the rest of our tables than to have a second set of read-only tables that live inside Nexus.  That's especially true if the database tables refer to entries in the in-memory tables.

One case where I can imagine an in-database table would actually differ from an in-memory table is if we add a new built-in role.  In that case, the first Nexus that we deploy with that role will insert it into the database.  Other Nexus instances will immediately see this role, e.g., if you were to list roles.  However, their Oso Policy files won't know about it, so I don't think they'll ever look for this role.  So this improvement is pretty marginal.

In summary, I'm still on the fence about it, but it seems to me slightly better than the alternative.  What do you think?

File: nexus/src/db/datastore.rs (hunk 2 above)
Comment (smklein) [reply nexus/src/db/datastore.rs:2168]: I have a working version of `paginated_multicolumn` that integrates here. Would you like the commits to be added to this PR, or uploaded as a separate PR?

Comment (davepacheco) [reply nexus/src/db/datastore.rs:2168]: Nice!  I'd love a separate PR just so I can better keep things straight, particularly with #520 also in flight.  Thanks for doing that!

Comment (davepacheco) [reply nexus/src/db/datastore.rs:2168]: (separate PR is #522)

File: nexus/tests/integration_tests/roles_builtin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/tests/integration_tests/roles_builtin.rs:72]: Fixed in f1775564.

File: common/src/sql/dbinit.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply common/src/sql/dbinit.sql]: Fixed in f1775564.

File: nexus/src/external_api/views.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/views.rs]: I took a stab at this in 77997282.  I think it's an improvement but it has some warts I'd love your opinion on.

Here's the issue:

* Within the database, the "resource_type" is a string.  This refers to some other resource table (e.g., "project") that the role applies to.  The short explanation for this being a string is that I didn't want to have to do a schema update in order to add a role to an object that doesn't already have a role.  I put much more detail below.
* For the newtype `RoleName`, I wanted it to be the case that an instance of `RoleName` represents a valid role name.  And I wanted to store the parsed representation, similar to `Name`.  I figured it makes sense to validate that the resource type _part_ of the name (the "project" in "project.viewer") corresponds to an actual `ResourceType` variant.
* This introduces a failure mode where you can load a `role_builtin` from the database whose name isn't a valid resource type.  This should only be possible due to a previous bug or else because we've rolled out a Nexus upgrade that delivers a new resource type with a new built-in role and an old Nexus handles this request.  The way this shows up is in the endpoint handler when we construct the view, it can fail.  We treat this as a 500.  It should be really uncommon unless there's a bug, in which case a 500 is correct.

There are a few other ways to deal with this:

1. Don't validate the resource_type of the RoleName when we parse it, which means storing it as a String rather than a ResourceType.  That's okay I guess, but it just means that code in Nexus that happens to _have_ a RoleName cannot rely on it corresponding to a valid resource.  (Maybe that's okay, since it already can't rely on it being a valid role?)  This approach also means that Nexus can return RoleNames for resource types it doesn't know about, which is a little weird, but admittedly seems correct in the upgrade case I mentioned above.
2. On the other end of the spectrum, store a ResourceType rather than a String in the `db::model::RoleBuiltin`.  The net result is basically the same as what we have today: if there's a resource type in the database that we don't know about, we're going to fail the request with an error when we try to deserialize it.  I don't like this because I prefer to let the consuming code decide if this is a problem or not.
3. Constrain the database values.  There are several ways to do _this_ too.  I don't like this approach because it means that adding built-in roles to a resource that didn't have them before requires a schema update -- that sucks.  Plus, it doesn't solve this problem: some other Nexus might have done such a schema update and now we're basically back in case 2.

As I write this, I've nearly talked myself into alternative 1.

As an aside, I've changed the serialized form of `ResourceType` to use underscores instead of spaces between words, so "vpc subnet" becomes "vpc_subnet".  I did this mainly because I didn't think it mattered much either way, and this is easier to programmatically serialize and deserialize because we can just use `style = "snake_case"`.  (I had originally used spaces because it's slightly more user-friendly, but I don't think it's worthwhile now.)

Finally, this isn't super relevant after all but I'd already written it, so here's some more context about the schema representation.

<details>
The problem is basically that roles (and role assignments, once we get to #520) need to refer to some other database table.  I will call this other table the "target" for lack of a better word. (So in the role "project.viewer", the target is the "project" table.)  In #520, we have actual role assignments, and we have an actual foreign key into that table.  I'm not aware of any great patterns for doing this in the database.  Here are some options:

- use a separate table for every possible target.  So instead of "roles_builtin", we'd have "roles_builtin_project", "roles_builtin_organization", etc.  Cons: lots of tables, lots of boilerplate, and adding roles on a new target (even if that resource already exists) requires a schema update.
-  create a custom datatype (an enum) and store that in the table.  So the "resource_type" column for "roles_builtin" is a custom enum.  Cons: adding roles on a new target (even if that resource already exists) requires a schema update.
- keep it as a string and have Nexus just know that it's a table name.  Cons: kind of implicit, and allows the database to become inconsistent or incorrect.  But huge pro: no schema change necessary (and often no code change necessary) to add roles on new targets.

I opted for this last option.
</details>

[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/views.rs]: Okay, after all that, I talked myself into not validating the resource type in RoleName.  It wouldn't validate that the name referred to a real role, anyhow.  Plus, nobody was using the typed version -- it's read-only at this point -- so it's not like somebody else needs to validate it anywhere else.  I made this change in a4524baf4a1e90c57208d7028c201443b195f165.

I kept the change to `ResourceType`'s serialized form because this issue made me realize that we probably _do_ want to make this change now, while we still can without breaking any consumers.  However, it's now unrelated to the rest of this change.

File: common/src/sql/dbinit.sql (hunk 3 above)
Comment (davepacheco) [reply common/src/sql/dbinit.sql:750]: FWIW, having fleshed out #520, I still like that this is a separate table.  Nothing's really changed but it feels more consistent with what I expect we'll wind up doing with custom roles, and at this point it would be more work to undo it without much obvious benefit.

File: common/src/api/external/mod.rs [hunk 7]
```diff
@@ -296,6 +296,85 @@ impl Name {
...
+    Eq,
+    FromStr,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    SerializeDisplay,
+)]
+#[display("{resource_type}.{role_name}")]
+pub struct RoleName {
+    #[from_str(regex = "[a-z_]+")]
```
Comment (teisenbe) [inline common/src/api/external/mod.rs:316]: Should we use - instead of _ to be consistent with name strings?

File: common/src/sql/dbinit.sql (hunk 3 above)
Comment (teisenbe) [reply common/src/sql/dbinit.sql:750]: Either is probably fine

File: nexus/src/external_api/views.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (teisenbe) [reply nexus/src/external_api/views.rs]: This seems fine. It might be good to put a comment in explaining that choice

File: common/src/api/external/mod.rs (hunk 7 above)
Comment (davepacheco) [reply common/src/api/external/mod.rs]: Done in 38fccbff.  I changed this in both the resource type and the role name.

File: nexus/src/external_api/views.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply nexus/src/external_api/views.rs]: Done in 38fccbff.

**Comment by davepacheco** [comment]
The latest changes address review feedback, sync with "main", and pull in #522.  I think I've addressed all the feedback, and #522 was reviewed and approved separately, so I'm planning to land this once tests pass.
